### DISCUSSION TITLE
React Hook Form Integration

### DISCUSSION THREAD

**Discussion by vankhoawin** [twilio-labs/paste#823] [Show and tell]
Code Sandbox: https://codesandbox.io/s/paste-starter-kit-with-react-hook-form-ibd8e?file=/src/Form.tsx

Hi team,

I played around with integrating the form library [React Hook Form](https://react-hook-form.com/) with Paste and I had pretty satisfying results. Because Paste lightly wraps and correct passes refs forward, it was pretty easy to integrate the two (versus [having to create declarative bindings for formik](https://github.com/twilio-labs/formik-paste)). I did have to add a little code to integrate RadioGroup since it relies on React Context, see code snippet below (truncated from the sandbox):

```tsx
      <Controller
        control={control}
        name="radio"
        render={({ onChange, value, name }) => (
          <RadioGroup
            name={name}
            legend={
              <Text as="span" color="currentColor">
                This is some help text with a{" "}
                <Anchor href="http://paste.twilio.com">link</Anchor>.
              </Text>
            }
            required
            orientation="horizontal"
            onChange={onChange}
            value={value}
          >
            <Radio
              id={inputID("radio1")}
              name={name}
              value="1"
              helpText={
                <Text as="span" color="currentColor">
                  This is some legend text with a{" "}
                  <Anchor href="http://paste.twilio.com">link</Anchor>.
                </Text>
              }
            >
              First
            </Radio>
          </RadioGroup>
        )}
      />
```

I guess my ask is more of, what are your thoughts on this integration? Do you feel like there's a way to decouple context from the RadioGroup, or if using React Hook Form's [supplied Controller component](https://react-hook-form.com/api#Controller) to accomplish this is fine as is?

**Comment by kate-porter** [comment]
Hey @vnguyen94! Forwarding this discussion along to eng - we'll be in touch soon!

**Comment by TheSisb** [comment]
Hi @vnguyen94,

It's my first time hearing about React Hook Form and I really like it so far! So I did some digging into the library and short answer: I think using the supplied Controller is the correct path forward.

Slightly longer answer: I was shocked that CheckboxGroup worked but RadioGroup did not, so I did some digging.  RadioGroup uses context but CheckboxGroup maps over children and clones them with additional props passed through. I think that CheckboxGroup has to be updated to using context too, so that component may need a controller as well.  Initially, RadioGroup did the children mapping too, until this [issue was reported](https://github.com/twilio-labs/paste/discussions/535).  As you can understand, I don't think there's a clean way to provide the Group functionality with override-able props if we map children rather than use React context. These Group components are necessary to Paste because it makes downstream style updates easier to manage in the future.  If you have any additional thoughts or suggestions please report back!

This is **really nice work** here! Thanks for bringing all this to our attention!

**Reply by vankhoawin** [reply]
Was about to share components with another team when I realized they would have to import in formik and formik-paste, so thought I'd try alternatives to see what else is out there (IoT uses it in their forms). Compared to formik so far, i've only seen less wrappers, smaller bundle size, and much better performance (our current large forms noticeably slow down when typing quickly).

> I think that CheckboxGroup has to be updated to using context too, so that component may need a controller as well.

Oh, should I wait until this is done before migrating over? Or if a ticket is created, can I be made a watcher?

And I remember now the other reason why I was asking: for FormInput and Select, I had to add a `@ts-ignore` label because they were both "missing" the `onChange` and `value` props.

**Reply by TheSisb** [reply]
> Compared to formik so far, i've only seen less wrappers, smaller bundle size, and much better performance (our current large forms noticeably slow down when typing quickly).

Good to know!  I made two tickets in our backlog and pinged you on both 👍 

You should be able to start your migration to this new library if you'd like to, just be aware that you should use the Controller on CheckboxGroup so it's future-proof.

**Reply by vankhoawin** [reply]
That's perfect. Thank you!

**Comment by TheSisb** [comment]
@vnguyen94 The CheckboxGroup changes are in a PR and will probably land today (https://github.com/twilio-labs/paste/pull/858). Just an FYI 👍

**Comment by richbachman** [comment]
Hi @vnguyen94 we just merged https://github.com/twilio-labs/paste/pull/1251. Once everything builds you should be able to use the various form components without needing to add `@ts-ignore`. Let us know if you find any other issues.

**Reply by vankhoawin** [reply]
Awesome, and will do
